daemons/based/pacemaker-based.c

```diff
@@ -52,141 +55,149 @@ static bool startCib(const char *filename);
+      stand_alone_cb, "(Advanced use only) Run in stand-alone mode", NULL },
+
+    { "disk-writes", 'w', G_OPTION_FLAG_NONE, G_OPTION_ARG_NONE,
+      &cib_writes_enabled,
+      "(Advanced use only) Enable disk writes (enabled by default unless in "
+      "stand-alone mode)", NULL },
+
+    { "cib-root", 'r', G_OPTION_FLAG_NONE, G_OPTION_ARG_STRING, &cib_root,
+      "(Advanced use only) Directory where the CIB XML file should be located "
+      "(default: " CRM_CONFIG_DIR ")", NULL },
+
+    { NULL }
 };
 
+static GOptionContext *
+build_arg_context(pcmk__common_args_t *args)
+{
+    GOptionContext *context = NULL;
+
+    context = pcmk__build_arg_context(args, NULL, NULL, "[metadata]");
+    pcmk__add_main_args(context, entries);
+    return context;
+}
+
 int
 main(int argc, char **argv)
 {
-    int flag;
-    int rc = 0;
-    int index = 0;
-    int argerr = 0;
-    struct passwd *pwentry = NULL;
     crm_ipc_t *old_instance = NULL;
 
+    GError *error = NULL;
+
+    pcmk__common_args_t *args = pcmk__new_common_args(SUMMARY);
+    gchar **processed_args = pcmk__cmdline_preproc(argv, NULL);
+    GOptionContext *context = build_arg_context(args);
+
     crm_log_preinit(NULL, argc, argv);
-    pcmk__set_cli_options(NULL, "[options]", long_options,
-                          "daemon for managing the configuration "
-                          "of a Pacemaker cluster");
 
-    mainloop_add_signal(SIGTERM, cib_shutdown);
-    mainloop_add_signal(SIGPIPE, cib_enable_writes);
+    if (!g_option_context_parse_strv(context, &processed_args, &error)) {
+        if (exit_code == CRM_EX_OK) {
+            exit_code = CRM_EX_USAGE;
+        }
+        goto done;
+    }
 
-    cib_writer = mainloop_add_trigger(G_PRIORITY_LOW, write_cib_contents, NULL);
+    if (args->version) {
+        g_strfreev(processed_args);
+        pcmk__free_arg_context(context);
 
-    while (1) {
-        flag = pcmk__next_cli_option(argc, argv, &index, NULL);
-        if (flag == -1)
-            break;
+        /* FIXME: When pacemaker-based is converted to use formatted output,
+         * this can go.
+         */
+        pcmk__cli_help('v', CRM_EX_OK);
+    }
 
-        switch (flag) {
-            case 'V':
-                crm_bump_log_level(argc, argv);
-                break;
-            case 's':
-                stand_alone = TRUE;
-                preserve_status = TRUE;
-                cib_writes_enabled = FALSE;
-
-                pwentry = getpwnam(CRM_DAEMON_USER);
-                CRM_CHECK(pwentry != NULL,
-                          crm_perror(LOG_ERR, "Invalid uid (%s) specified", CRM_DAEMON_USER);
-                          return CRM_EX_FATAL);
-
-                rc = setgid(pwentry->pw_gid);
-                if (rc < 0) {
-                    crm_perror(LOG_ERR, "Could not set group to %d", pwentry->pw_gid);
-                    return CRM_EX_FATAL;
-                }
+    mainloop_add_signal(SIGTERM, cib_shutdown);
```

Comment:
I haven't looked closely yet (out of scope for this PR), but I wonder if these calls even work right before the mainloop has been initialized.